Document what Nix *is* (#6420)
NixOS/nix: doc/manual/src/design/store/objects.md

# Store Objects

Data in Nix is chiefly organized into *store objects*.
A store object is the pair of

>   - A (root) file system object
>   - A set of references to store objects

Reviewer: ```suggestion
  - a file system object
  - a set of references to store objects.
```

It's not clear what "root" would mean here, because we have not yet defined what a file system object is, and the root notion only explained at the end.

## File system objects

> The Nix store uses a simple filesystem model.

Reviewer: ```suggestion
The Nix store uses a simple file system model.
```

>     data FileSystemObject
>       = Regular Executable ByteString
>       | Directory (Map FileName FSO)
>       | SymLink ByteString
> 
>     data Executable
>       = Executable
>       | NonExecutable

Reviewer: ```suggestion
    data FileSystemObject
      = File Executable Data
      | Directory (Map FileName FileSystemObject)
      | SymLink Path
```

I assume you've taken this from the thesis. Suggestion: Make more self-explanatory names, and skip the details for brevity.

> In particular, every file system object falls into these three cases:
> 
>  - File: an executable flag, and arbitrary data
> 
>  - Directory: mapping of names to child file system objects.
> 
>  - Symlink: may point anywhere.

Reviewer: ```suggestion
Every file system object is one of the following:
 - File: an executable flag, and arbitrary data
 - Directory: mapping of names to child file system objects
 - [Symbolic link](https://en.m.wikipedia.org/wiki/Symbolic_link): may point anywhere.
```
```

>    In particular, symlinks that do not point within the containing root file system object or that of another store object referenced by the containing store object are allowed, but might not function as intended.

Reviewer: ```suggestion
   In particular, symlinks pointing to outside of their own root file system object or to store objects not referenced by their root are allowed, but might not function as intended.
```

picking up https://github.com/NixOS/nix/pull/4280/files#r832303802

> A bare file or symlink as the "root" file system object is allowed.

Reviewer: ```suggestion
A bare file or symlink as the root file system object is allowed.
```

### Comparison with Git

This is close to Git's model, but with one crucial difference:
Git puts the "permission" info within the directory map's values instead of making it part of the file (blob, in it's parlance) object.

    data GitObject
      = Blob ByteString
      | Tree (Map FileName (Persission, FSO))

    data Persission
      = Directory -- IFF paired with tree
      -- Iff paired with blob, one of:
      | RegFile
      | ExecutableFile
      | Symlink

So long as the root object is a directory, the representations are isomorphic.
There is no "wiggle room" the git way since whenever the permission info wouldn't matter (e.g. the child object being mapped to is a directory), the permission info must be a sentinel value.

However, if the root object is a blob, there is loss of fidelity.
Since the permission info is used to distinguish executable files, non-executable files, and symlinks, but there isn't a "parent" directory of the root to contain that info, these 3 cases cannot be distinguished.

> Git's model matches Unix tradition, but Nix's model is more natural.

Reviewer: This is interesting to know, but it diverges from our topic and therefore should be somewhere else. Maybe move that to advanced topics for now. Eventually there should be a section on relating Nix to other tools, such as Git, IPFS, Bazel, etc.

## References

> Store objects can refer to both other store objects and themselves.

Reviewer: ```suggestion
A store object can refer to both other store objects and itself.
```
